src/devices/Mfrc522/Mfrc522Controller.cs

```diff
@@ -0,0 +1,447 @@
+            var (status, backData, backBits) = SendCommand(Transceive, buf.ToArray());
+
+            if (status != OK || backBits != 0x18)
+                return 0;
+
+            Debug.WriteLine($"Size: {backData[0]}");
+            return backData[0];
+
+        }
+
+        public Status Authenticate(RequestMode authenticationMode, byte blockAddress, byte[] sectorKey, byte[] serialNumber)
+        {
+            // First byte should be the authMode (A or B) Second byte is the trailerBlock (usually 7)
+            var buff = new List<byte> { (byte)authenticationMode, blockAddress };
+
+            // Now we need to append the authKey which usually is 6 bytes of 0xFF
+            var i = 0;
+            while (i < sectorKey.Length)
+            {
+                buff.Add(sectorKey[i]);
+                i++;
+            }
+
+            i = 0;
+
+            // Next we append the first 4 bytes of the UID
+            while (i < 4)
+            {
+                buff.Add(serialNumber[i]);
+                i++;
+            }
+
+            // Now we start the authentication itself
+            var (status, _, _) = SendCommand(Command.Authenticate, buff.ToArray());
+
+            // Check if an error occurred
+            if (status != OK)
+            {
+                throw new Mfrc522Exception("AUTH ERROR!!");
+            }
+
+            if ((ReadSpi(Status2Reg) & 0x08) == 0)
+            {
+                throw new Mfrc522Exception("AUTH ERROR(status2reg & 0x08) != 0");
+            }
+
+            // Return the status
+            return status;
+        }
+
+        public void ClearSelection()
+        {
+            ClearBitMask(Status2Reg, 0x08);
+        }
+
+        public void WriteCardData(byte blockAddress, byte[] writeData)
+        {
+            var buff = new List<byte> { (byte)RequestMode.Write, blockAddress };
+            var crc = CalulateCRC(buff.ToArray());
+            buff.Add(crc[0]);
+            buff.Add(crc[1]);
+            var (status, backData, backLen) = SendCommand(Transceive, buff.ToArray());
+
+            if (status != OK || backLen != 4 || (backData[0] & 0x0F) != 0x0A)
+            {
+                status = Error;
+            }
+
+            Debug.WriteLine($"{backLen} backdata &0x0F == 0x0A {backData[0] & 0x0F}");
+
+            if (status != OK) return;
+
+            var i = 0;
+            var buf = new List<byte>();
+            while (i < 16)
+            {
+                buf.Add(writeData[i]);
+                i++;
+            }
+
+            crc = CalulateCRC(buf.ToArray());
+            buf.Add(crc[0]);
+            buf.Add(crc[1]);
+
+            (status, backData, backLen) = SendCommand(Transceive, buf.ToArray());
+
+            if (status != OK || backLen != 4 || (backData[0] & 0x0F) != 0x0A)
+            {
+                throw new Mfrc522Exception("Failed to write to");
+            }
+        }
+
+        public void DumpClassic1K(byte[] key, byte[] uid)
```

Comment:
Please use Span/ReadOnlySpan as arguments, also might be worth expanding the name a bit - I presume this refers to Mifare Classic?